h2o-py/h2o/explanation/_explain.py

```diff
@@ -2209,16 +2286,25 @@ def _process_models_input(
     :returns: tuple (is_aml, models_to_show, classification, multinomial_classification,
                     multiple_models, targets, tree_models_to_show)
     """
-    is_aml = isinstance(models, h2o.automl._base.H2OAutoMLBaseMixin)
-    if is_aml:
-        models_to_show = [models.leader]
-        multiple_models = models.leaderboard.nrow > 1
+    from itertools import islice
+    is_aml = False
+    if _is_automl_or_leaderboard(models):
+        is_aml = True
+        models = models.leaderboard if isinstance(models, h2o.automl._base.H2OAutoMLBaseMixin) else models
+        models_to_show = [h2o.get_model(models[0, "model_id"])]
+        multiple_models = models.nrow > 1
+        models_with_varimp = models_to_show if _has_varimp(models_to_show[0]) \
+            else list(islice((h2o.get_model(model)
+                              for model in _get_model_ids_from_automl_or_leaderboard(models, filter_=_has_varimp)
+                              if _has_varimp(h2o.get_model(model))), 1))
```

Comment:
it's very inefficient to call `h2o.get_model` twice or thrice here (filter=_has_varimp will call get_model), each call makes a request to the backend.
Also, I don't understand why `_has_varimp` appears both in the `filter_` and in the `if` guard.
Finally, usage of `list(islice(..., 1))` is rather unorthodox, you probably just need `next(...)`.
Overall, this piece can be improved for both readability and efficiency.

Reply:
The `_get_model_ids_from_automl_or_leaderboard(models, filter_=_has_varimp)` will return model ids of all algos but StackedEnsembles and NaiveBayes (they don't have varimp). 

The ` if _has_varimp(h2o.get_model(model))` is there to check that the model that should have varimp really has it - sometimes xgboost didn't for some reason.

I chose `list(islice(..., 1))` because of the `next([])` behavior - I'd have to surround it with `try-except` block. 

I will split it in to multiple lines, it will save one call to `h2o.get_model` and it should improve readability as well.

Reply:
`next(comprehension, def_val)` doesn't raise exception.
You may need 2 steps, `next` followed with a `if-else` clause but Python has a strong focus on readability, let's use it.

> The if _has_varimp(h2o.get_model(model)) is there to check that the model that should have varimp really has it - sometimes xgboost didn't for some reason.

it still doesn't make much sense to me: you're using the same predicate function in both cases, so why in the first case it could return a "model that should have varimp [but maybe doesn't have it]", whereas the second case ensures that "the model that should have varimp really has it".

Reply:
> it still doesn't make much sense to me: you're using the same predicate function in both cases, so why in the first case it could return a "model that should have varimp [but maybe doesn't have it]", whereas the second case ensures that "the model that should have varimp really has it".

The first application  of the `_has_varimp` checks only on the model_id - so it's basically just a coarse filter to get rid of the SE and NaiveBayes. The second application is on the "materialized" model so it can be sure that it does have the varimp. 

I see I should really improve readability and possibly split the has_varimp in to two functions one for model_ids and one for models.

Thank you for mentioning `next(..., default)`, I didn't know that.

Reply:
> I see I should really improve readability and possibly split the has_varimp in to two functions one for model_ids and one for models.

exactly, you can't have this function have a different semantic depending if the model parameter is passed as a model or as a model_id